File: app/src/main/java/com/sk/revisit/components/UrlBarComponent.java
package com.sk.revisit.components;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.Filter;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.AppCompatAutoCompleteTextView;

import com.sk.revisit.activities.BaseActivity;
import com.sk.revisit.helper.FileHelper;
import com.sk.revisit.webview.MyWebView;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.Executor;
import java.util.concurrent.Executors;

public class UrlBarComponent extends Component {
    private final AppCompatAutoCompleteTextView urlAppCompatAutoCompleteTextView;
    private final MyWebView mainWebView;
    String rootPath;

    public UrlBarComponent(BaseActivity activity, AppCompatAutoCompleteTextView mUrlAppCompatAutoCompleteTextView, MyWebView mainWebView, String rootPath) {
        super(activity);
        urlAppCompatAutoCompleteTextView = mUrlAppCompatAutoCompleteTextView;
        this.mainWebView = mainWebView;
        this.rootPath = rootPath;
        init();
    }

    void init() {
        //init urlAppCompatAutoCompleteTextView
        urlAppCompatAutoCompleteTextView.setOnFocusChangeListener((view, hasFocus) -> {
            if (hasFocus) return;
            try {
                mainWebView.loadUrl(urlAppCompatAutoCompleteTextView.getText().toString());
            } catch (Exception e) {
                alert(e.toString());
            }
        });

        urlAppCompatAutoCompleteTextView.setOnEditorActionListener((v, actionId, event) -> {
            try {
                mainWebView.loadUrl(urlAppCompatAutoCompleteTextView.getText().toString());
            } catch (Exception e) {
                alert(e.toString());
            }
            return true;
        });

        setupAutoComplete(new File(rootPath));
    }

    public void setupAutoComplete(File directory) {
        Executor executor = Executors.newFixedThreadPool(4);
        CompletableFuture.supplyAsync(() -> {
            //List<String> fileNames = new ArrayList<>();
            //FileHelper.searchRecursive(directory, ".html",fileNames);
            try {
                return FileHelper.search(directory, ".html");
            } catch (IOException e) {
                activity.alert(e.toString());
                return null;
            }
        }, executor).thenAccept(fileNames -> {
            ArrayAdapter<String> adapter = new CustomCompletionAdapter(activity,android.R.layout.simple_dropdown_item_1line, fileNames);
            activity.runOnUiThread(() -> urlAppCompatAutoCompleteTextView.setAdapter(adapter));
        });
    }

    public void setText(String text) {
        activity.runOnUiThread(() -> urlAppCompatAutoCompleteTextView.setText(text));
    }
}

class CustomCompletionAdapter extends ArrayAdapter<String> {

    private List<String> org, curr;

    public CustomCompletionAdapter(@NonNull Context context, int resource, List<String> data) {
        super(context, resource);
        org = new ArrayList<>(data);
        curr = new ArrayList<>(data);
    }

    @NonNull
    @Override
    public View getView(int position, @Nullable View convertView, @NonNull ViewGroup parent) {
        if (convertView == null) {
            convertView = LayoutInflater.from(getContext()).inflate(android.R.layout.simple_dropdown_item_1line, parent, false);
        }
        TextView textView = convertView.findViewById(android.R.id.text1);
        textView.setText(getItem(position));
        return convertView;
    }

    @NonNull
    @Override
    public Filter getFilter() {
        return new Filter() {
            @Override
            protected FilterResults performFiltering(CharSequence charSequence) {
                FilterResults results = new FilterResults();
                if (charSequence != null) {
                    String searchText = charSequence.toString().toLowerCase();
                    List<String> filteredList = new ArrayList<>();
                    for (String item : org) {
                        if (item.toLowerCase().contains(searchText)) {
                            filteredList.add(item);
                        }
                    }
                    results.values = filteredList;
                    results.count = filteredList.size();
                }
                return results;
            }

            @Override
            protected void publishResults(CharSequence charSequence, FilterResults filterResults) {
                curr.clear();
                if (filterResults != null && filterResults.count > 0) {
                    curr.addAll((List<String>) filterResults.values);
                    notifyDataSetChanged();
                } else {
                    curr.addAll(org);
                    notifyDataSetChanged();
                }
            }
        };
    }

    @Nullable
    @Override
    public String getItem(int position) {
        return curr.get(position);
    }

    @Override
    public int getCount() {
        return curr.size();
    }
}

File: app/src/main/java/com/sk/revisit/components/UrlBarComponent.kt
package com.sk.revisit.components

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.Filter
import android.widget.TextView
import androidx.appcompat.widget.AppCompatAutoCompleteTextView
import com.sk.revisit.activities.BaseActivity
import com.sk.revisit.helper.FileHelper
import com.sk.revisit.webview.MyWebView
import kotlinx.coroutines.*
import java.io.File
import java.io.IOException

class UrlBarComponent(
    activity: BaseActivity,
    private val urlAutoCompleteTextView: AppCompatAutoCompleteTextView,
    private val mainWebView: MyWebView,
    private val rootPath: String
) : Component(activity) {

    // Coroutine scope for background file searching
    private val componentScope = CoroutineScope(Dispatchers.Main + SupervisorJob())

    init {
        setupUrlBarListeners()
        setupAutoComplete(File(rootPath))
    }

    private fun setupUrlBarListeners() {
        urlAutoCompleteTextView.setOnFocusChangeListener { _, hasFocus ->
            if (!hasFocus) loadUrlFromInput()
        }

        urlAutoCompleteTextView.setOnEditorActionListener { _, _, _ ->
            loadUrlFromInput()
            true
        }
    }

    private fun loadUrlFromInput() {
        val url = urlAutoCompleteTextView.text.toString().trim()
        if (url.isNotEmpty()) {
            try {
                mainWebView.loadUrl(url)
            } catch (e: Exception) {
                alert("Error loading URL: ${e.message}")
            }
        }
    }

    fun setupAutoComplete(directory: File) {
        componentScope.launch {
            val fileNames = withContext(Dispatchers.IO) {
                try {
                    FileHelper.search(directory, ".html")
                } catch (e: IOException) {
                    withContext(Dispatchers.Main) { alert("Error searching files: ${e.message}") }
                    emptyList<String>()
                }
            }

            if (fileNames.isNotEmpty()) {
                urlAutoCompleteTextView.setAdapter(CustomCompletionAdapter(activity, fileNames))
            }
        }
    }

    fun setText(text: String) {
        activity.runOnUiThread { urlAutoCompleteTextView.setText(text) }
    }

    fun release() {
        componentScope.cancel() // Cancels any ongoing background searches
    }
}

class CustomCompletionAdapter(context: Context, data: List<String>) :
    ArrayAdapter<String>(context, android.R.layout.simple_dropdown_item_1line, data) {

    private val originalList = data.toList()
    private var filteredList = data.toList()

    override fun getCount() = filteredList.size
    override fun getItem(position: Int) = filteredList.getOrNull(position)

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        val view = convertView ?: LayoutInflater.from(context)
            .inflate(android.R.layout.simple_dropdown_item_1line, parent, false)

        view.findViewById<TextView>(android.R.id.text1).text = getItem(position)
        return view
    }

    override fun getFilter(): Filter = object : Filter() {
        override fun performFiltering(constraint: CharSequence?): FilterResults {
            val results = FilterResults()
            val filtered = if (constraint.isNullOrEmpty()) {
                originalList
            } else {
                val query = constraint.toString().lowercase()
                originalList.filter { it.lowercase().contains(query) }
            }
            results.values = filtered
            results.count = filtered.size
            return results
        }

        @Suppress("UNCHECKED_CAST")
        override fun publishResults(constraint: CharSequence?, results: FilterResults?) {
            filteredList = results?.values as? List<String> ?: originalList
            notifyDataSetChanged()
        }
    }
}